### PR TITLE
Add KIVA_WERROR option to interpret compiler warnings as errors

### PR DESCRIPTION
**Pull request by galanCA**
## Description

The goal is to have the option to turn warnings into errors by 

### Major changes:

- Add compiler-flag.cmake and an interface library to distribute the compiler flags
- Add the option to turn off and on Werror (/WX for windows).

### CODE REVIEW

**Comment by galanCA** [comment]
@nealkruis 
- What should the default option for `KIVA_WERROR` be?
- Would you like the CI to have Werror ON or OFF?

**Comment by nealkruis** [comment]
I go back and forth on the default. I think I want it on by default, and then executables linking to Kiva can decide if they want to turn it off.

Definitely needs to be on for the CI, otherwise it won't keep us from merging warnings.

File: cmake/compiler-flags.cmake [hunk 1]
```diff
@@ -0,0 +1,31 @@
+add_library(kiva_common_interface INTERFACE)
+
+  #================#
+  # Compiler flags #
+  #================#
+
+target_compile_options(kiva_common_interface INTERFACE
```
Comment (nealkruis) [inline cmake/compiler-flags.cmake:7]: I'm having second thoughts about this. Making these `INTERFACE` options means that they will propagate all the way up to the linking executable whether it wants them or not. Can you come up with a DRY way to apply these as `PRIVATE` options on the individual targets?

File: CMakeLists.txt [hunk 2]
```diff
@@ -8,6 +8,7 @@ option(ENABLE_OPENMP "Use OpenMP" OFF)
 option( KIVA_GROUND_PLOT "Build ground plotting library" ON )
 mark_as_advanced(FORCE BUILD_GROUND_PLOT)
 option( KIVA_TESTING "Build tests" ON )
+option( KIVA_WERROR "Turn on warnings into error" ON )
```
Comment (nealkruis) [inline CMakeLists.txt]: I changed my mind on this one. Let's make it default to OFF, and handle things similarly to the [fmt repository](https://github.com/fmtlib/fmt).

**Review by nealkruis — CHANGES_REQUESTED** [review]

File: cmake/compiler-flags.cmake (hunk 1 above)
Comment (nealkruis) [reply cmake/compiler-flags.cmake:7]: Actually...I think you can leave this how it is, but make sure the targets that link to it set it as `PRIVATE`.

File: cmake/compiler-flags.cmake [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+>
+
+$<$<CXX_COMPILER_ID:Clang>:
+  -Wall 
+  -Wextra 
+  -Wpedantic
+  $<$<BOOL:${KIVA_WERROR}>:
+    -Werror # Turn warnings into errors
+  >
+>
```
Comment (nealkruis) [inline cmake/compiler-flags.cmake]: This is WET.

**Comment by galanCA** [comment]
@nealkruis I tested that the interface library was only applied to the corresponding target by:
- Creating a new branch in CSE where:
  - I checkout this branch for kiva
  - make sure to turn on the WX flag in Kiva
  - and turn CSE warning flag from W3 to W4
From this, I would expect the CSE warnings to turn into errors if the WX flags were passed to CSE. But I was able to see the warnings and no errors allowing CSE to build successfully.

I went back and added the ability to turn on `KIVA_WERROR` option inside `vendor/cmakelist.txt` and check the `CMakeCache.txt`. But the output was the same as before, I could see the warnings, no errors and CSE build successfully. From this, I derived that the interface library is only affecting the private target and not the dependencies of the target. 

Do you agree with this conclusion?

File: src/libgroundplot/CMakeLists.txt [hunk 4]
```diff
@@ -5,16 +5,10 @@ set(groundplot_src  GroundPlot.cpp
 
 add_library(groundplot STATIC ${groundplot_src})
 
-target_link_libraries(groundplot fmt mgl-static libkiva)
+target_link_libraries(groundplot fmt mgl-static libkiva kiva_common_interface)
```
Comment (nealkruis) [inline src/libgroundplot/CMakeLists.txt]: No keyword?

File: src/kiva/CMakeLists.txt [hunk 5]
```diff
@@ -12,16 +12,10 @@ set(CMAKE_INSTALL_RPATH "@executable_path")
 
 add_executable(kiva ${kiva_src})
 
-target_link_libraries(kiva libkiva groundplot boost_program_options yaml-cpp)
+target_link_libraries(kiva libkiva groundplot boost_program_options yaml-cpp kiva_common_interface)
```
Comment (nealkruis) [inline src/kiva/CMakeLists.txt]: No keyword?

**Review by nealkruis — CHANGES_REQUESTED** [review]
Getting very close. I left two comments, but I think we are also missing a link to the common interface on one of the targets.

**Review by nealkruis — APPROVED** [review]

**Pull request merged**